rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,87 @@
+
+* [Block pruning](https://github.com/tendermint/tendermint/issues/3652): removes historical blocks and associated data (e.g. validator sets) up to some height, keeping only the most recent blocks.
+
+* [State sync](https://github.com/tendermint/tendermint/issues/828): bootstraps a new node by syncing the state machine at a given height, but not historical blocks and associated data.
+
+To maintain the integrity of the chain, the use of these features must be coordinated such that necessary historical blocks will not be made unavailable or lost forever. In particular:
+
+* Some nodes should have complete block histories, for auditability and querying.
+
+* The majority of nodes should retain blocks longer than the Cosmos SDK unbonding period, for light client verification.
+
+* Some nodes must take and serve state sync snapshots with snapshot intervals less than the block retention periods, to allow new nodes to state sync and then replay blocks to catch up.
+
+* Only a minority of nodes can be state synced within the unbonding period, for light client verification and to serve block histories for catch-up.
+
+However, it is unclear if and how we should enforce this. It may not be possible to technically enforce all of these without knowing the state of the entire network, but it may also be unrealistic to expect this to be enforced entirely through social consensus. This is especially unfortunate since the consequences of misconfiguration can be permanent chain-wide data loss.
+
+The main configuration options involved are:
+
+* Unbonding time (SDK): the duration in which validators can be economically punished for misbehavior.
+
+* Block time (Tendermint): the minimum duration between consecutive blocks.
+
+* Block retention (Tendermint): the number of recent blocks (heights) to retain.
+
+* Snapshot interval (SDK): the interval (in heights) between taking state sync snapshots.
+
+* Snapshot retention (SDK): the number of recent state sync snapshots to retain.
+
+## Proposal
+
+* Unbonding time (SDK): already implemented as a genesis parameter `app_state.staking.params.unbonding_time`. Must now satisfy `unbonding_time < 2 * block.retention * block.time_iota_ms` otherwise errors (factor of 2 since `block.time_iota_ms` does not take into account block execution time).
+
+* Block time (Tendermint): already implemented as a genesis parameter `consensus_params.block.time_iota_ms`.
+
+* Block retention (Tendermint): implement as a new genesis parameter `consensus_params.block.retention` (default 0, i.e. retain all), and add a local configuration option `consensus.prune_blocks` (default off) that allows node operators to either enable or disable block pruning on a per-node basis. Must be at minimum 2, since the current and previous blocks are required for progress.
```

Comment:
Why does retention need to be a consensus parameter?

Reply:
It initial wasn't, but @melekes and @marbar3778 were concerned that if this was up to individual node operators then the chance of misconfiguration was high. For example, most node operators would probably choose a low retention on their own, which could lead to no-one running archival nodes and thus loss of historical blocks and inability to replay history. The idea was that if it is a consensus parameter then it would force operators to have a deliberate approach to it.

Personally I'm fine with either option, we just have to pick something.

Reply:
Another argument here is that if it's a consensus parameter then the SDK has access to it and can make sure the retention time is longer than the unbonding period. I don't believe this would currently be possible with a local Tendermint config option.

Reply:
With the former argument, I could argue that sane defaults and proper documentation should avoid situations of low-to-zero data availability.

The latter argument makes sense. But where is that enforced and checked? In the state machine? Then wouldn't that introduce a direct dependency on Tendermint? I suppose this is OK as we kind of already do this.

I really don't have a strong opinion here 👍

Reply:
> The latter argument makes sense. But where is that enforced and checked? In the state machine? Then wouldn't that introduce a direct dependency on Tendermint?

ABCI already provides consensus parameters to the application in `InitChain`, which would include the block retention - the SDK could then verify this against the unbonding time. After chain initialization, all consensus parameter changes must be done by the ABCI application itself.